### PR TITLE
Big rework of IR representation

### PR DESCRIPTION
**Pull request by ezyang**
You should read the commits individually.

### CODE REVIEW

File: torch/csrc/autograd/functions/jit_closure.cpp [hunk 1]
```diff
@@ -120,8 +120,8 @@ struct EmitNull : public Function {
 struct LambdaFunction : public Function {
   LambdaFunction(int num_inputs, std::function<variable_list(const variable_list&)> fn)
     : fn(fn) {
-    is_executable = true;
-    num_inputs = num_inputs;
+    this->is_executable = true;
+    this->num_inputs = num_inputs;
```
Comment (apaszke) [inline torch/csrc/autograd/functions/jit_closure.cpp]: I can't see why would that be necessary

File: setup.py [hunk 2]
```diff
@@ -359,6 +359,7 @@ def check_file(f):
     "torch/csrc/jit/tracer.cpp",
     "torch/csrc/jit/python_tracer.cpp",
     "torch/csrc/jit/interned_strings.cpp",
+    "torch/csrc/jit/type.cpp",
```
Comment (apaszke) [inline setup.py:362]: What's the problem with having it in `ir.h`?

File: torch/csrc/autograd/functions/jit_closure.cpp (hunk 1 above)
Comment (ezyang) [reply torch/csrc/autograd/functions/jit_closure.cpp]: num_inputs = num_inputs raises a clang warning that a value of int type is being assigned to itself. I can dig up the C++ standard if you like.

File: setup.py (hunk 2 above)
Comment (ezyang) [reply setup.py:362]: Type has no references to Graph/Node, and I'd like to keep it this way. The way to do that is move it to another header.

File: torch/csrc/autograd/functions/jit_closure.cpp (hunk 1 above)
Comment (ezyang) [reply torch/csrc/autograd/functions/jit_closure.cpp]: Yes, so if you bring a local num_inputs into scope, that will shadow the member variable. So `num_inputs = num_inputs` is always a no-op.

This is contrast to an initializer list in a constructor, e.g., the idiom `input_list(input_list)`, where the lhs reference is unambiguous since you can't initialize a local variable; it must be a member variable.

Comment (apaszke) [reply torch/csrc/autograd/functions/jit_closure.cpp]: ah lol. I only looked at `is_executable`. Nice catch

**Comment by zdevito** [comment]
Let's talk about these at the standup today because some of these commits go beyond refactoring.

File: torch/csrc/jit/linked_list.h [hunk 3]
```diff
@@ -0,0 +1,142 @@
+#include "torch/csrc/jit/assert.h"
+
+namespace torch { namespace jit {
```
Comment (zdevito) [inline torch/csrc/jit/linked_list.h]: Like we talked about in person, this should still be named graph_node_list. It is fine to keep it in its own file, but don't hide that we expect behavior like `destroy`

File: torch/csrc/jit/ir.h [hunk 4]
```diff
@@ -541,7 +537,19 @@ friend struct Node;
   }
 
   Node * create(NodeKind kind) {
-    return new Node(this,kind);
+    // NB: Node constructor adds node to all_nodes
+    Node * p = new Node(this, kind);
```
Comment (zdevito) [inline torch/csrc/jit/ir.h]: Just recapping: create shouldn't add inputs, we can have an `addInput(Node*)` variant for the case where we want to clone an input.    Return should also not have special create behavior. Asserts to check that extra return nodes are not being allocated are a good idea.

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 5]
```diff
@@ -196,30 +196,32 @@ struct GraphFuser {
...
-      size_t j = 0;
-      Node * new_output = graph->createClone(producer_for_chunk,[&](Node * n) {
-        auto & c = chunks[j++];
-        Node * ns = graph->createSelect(c,i);
-        ns->setType(sel->typeOption());
-        insertAfter(ns,c);
+      Node * sel = s.user;
+      JIT_ASSERT(sel->kind() == kSelect);
+      Node * new_output = graph->createClone(producer_for_chunk, [&](Node * input) {
+        Node * ns = graph->createClone(sel, [&](Node * chunk2) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/graph_fuser.cpp]: This isn't really a clone and the logic here to try to use clone makes it more complicated than before. You have a 2 nested lambdas, and a new environment. Instead, can we just add a copyMetadata  (type, stage, debugName) function?

File: torch/csrc/autograd/python_function.cpp [hunk 6]
```diff
@@ -502,9 +502,11 @@ static void _save_variables(THPFunction* self, t2var_type &t2var)
 }
 
 // t2var maps input and output tensors to variables
-static void _join_version_counters(THPFunction *self, t2var_type &t2var)
+static std::vector<std::pair<Variable, Variable>> // <input, output>
+_join_version_counters(THPFunction *self, t2var_type &t2var)
```
Comment (zdevito) [inline torch/csrc/autograd/python_function.cpp]: I am concerned that this approach is not safe, and that recording alias information is not the right way to go. The problem is that users won't know the aliasing status of the tensors going into a trace -- it is not part of the public API like sizes are, and not something people typically think about. If a trace is reused later with tensors 
 that have more aliasing than original inputs then in-place optimizations may then clobber data. We should discuss more about what alias information should be recorded. For instance, if the system does a semantic analysis to tag nodes with {no-alias, may alias, must alias} that might be sufficient for most optimizations. For instance Relu almost always follows conv or add, and we can simply mark conv or add no-alias and everything else may-alias. Then Relu can be marked in place.

File: torch/csrc/jit/ir.h [hunk 7]
```diff
@@ -58,6 +59,58 @@ static inline bool operator==(const Use & a, const Use & b) {
...
+    JIT_ASSERT(kind == AliasKind::NoneAlias || kind == AliasKind::MultiAlias);
+  }
+  Alias(AliasKind kind, size_t input) : kind_(kind), input_(input) {
+    JIT_ASSERT(input != std::numeric_limits<size_t>::max());
+    JIT_ASSERT(kind == AliasKind::SelectAlias || kind == AliasKind::DirectAlias);
+  }
+};
+
+// Can't use the GENERIC macros because alias is a value type
+#define ALIAS_IF(x, Kind) \
```
Comment (zdevito) [inline torch/csrc/jit/ir.h]: We can use switch statements as we discussed earlier.

**Review by zdevito — COMMENTED** [review]
Refactor commits are good, assuming the changes we discussed are made. 

Alias information warrants more discussion. I still think there is potentially a much simpler solution based on semantic analysis of the graph that would easily capture all places where people wrote in-place while also detecting in-place opportunities not expressed by explicitly.

File: torch/csrc/autograd/python_function.cpp (hunk 6 above)
Comment (ezyang) [reply torch/csrc/autograd/python_function.cpp]: This is a nuanced issue.

First, there are no cases where shared_pairs is dependent on the initial aliasing of the inputs to an autograd function. This can be determined by grepping for `mark_shared_storage` and noting that none of the invocations of this method from autograd functions depend on aliasing of inputs. (In one case, Resize, it is dependent on the *contiguity* of the input in question. Which is a good indication that traces should be indexed by the contiguity of their inputs; I don't believe we do this at the moment.) So, while the aliasing set (semantically, the set of nodes which alias one another) will vary depending on the aliasing of the initial inputs, the aliasing properties of operators is stable no matter the aliasing of the inputs.

You do raise a good point that if we eventually do an analysis which determines what inputs alias one another, this must be recomputed depended on the aliasing status; and in particular, if we want to insert data dependencies, we do eventually need to be mindful of the aliasing of the inputs. I don't think this can be conveniently done in the current API, since autograd closure generation is done without reference to the inputs (although, once again, if we index autograd closures by the aliasing of the initial inputs, this problem disappears.) It's also worth noting that the aliasing of inputs to the trace is unlikely to change across iterations of the loop, because aliasing is ~never data dependent.

Finally, the original intention of the alias intention was not to allow us to optimize operations into inplace (although, it is probably something profitable we can do), but to make sure we don't execute operations *explicitly* marked as inplace too early.

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 5 above)
Comment (ezyang) [reply torch/csrc/jit/passes/graph_fuser.cpp]: I think we should refactor this some more, but one question I have is, aren't *both* clones inappropriate? When I refactored this, the code smell I saw was that we had to fix the type of the clone of producer_for_chunk. It seems to me that in both cases, we should construct the things from scratch. (And thus I am not sure copyMetadata is appropriate here.)

**Comment by ezyang** [comment]
CR commits are labeled FIXUP, and are right after the commits they would be merged into.

File: torch/csrc/autograd/python_function.cpp [hunk 8]
```diff
@@ -500,9 +501,12 @@ static void _save_variables(THPFunction* self, t2var_type &t2var)
   self->to_save = NULL;
 }
 
-static void _join_version_counters(THPFunction *self, t2var_type &t2var)
+// t2var maps input and output tensors to variables
+static std::vector<std::pair<Variable, Variable>> // <input, output>
+_join_version_counters(THPFunction *self, t2var_type &t2var)
 {
-  if (!self->shared_pairs) return;
+  std::vector<std::pair<Variable, Variable>> return_pairs;
```
Comment (apaszke) [inline torch/csrc/autograd/python_function.cpp]: maybe make those `Variable*`? There's a lot of unnecessary refcounting going on because of this vector

File: torch/csrc/jit/attributes.h [hunk 9]
```diff
@@ -144,17 +144,28 @@ struct Attributes {
...
+  typename T::ValueType & get(Symbol name) const {
     auto it = find(name, true);
     T* child = dynamic_cast<T*>(it->get());
     JIT_ASSERT(child != nullptr);
     return child->value();
   }
   using AVPtr = AttributeValue::Ptr;
+  // NB: For determinism, we use a vector rather than a hash map.  This does
+  // mean that lookups are O(n), so you shouldn't use Attributes to store
+  // a big pile of messages.
```
Comment (apaszke) [inline torch/csrc/jit/attributes.h:156]: Why is that needed?

File: torch/csrc/jit/ir.h [hunk 10]
```diff
@@ -169,6 +59,46 @@ static inline bool operator==(const Use & a, const Use & b) {
...
+  // on multi-return nodes.  It's all equivalent under the
+  // Select Invariant).
+  Multi,
+  // This is a select node, and it aliases with the nth
+  // input of the node it is selecting from.
+  Select,
+  // This is not a multi-return node, and its output aliases
+  // with the nth input.
+  Direct,
+};
```
Comment (apaszke) [inline torch/csrc/jit/ir.h]: Is there any reason why we don't want to maintain aliasing information in the type?

File: torch/csrc/jit/ir.h [hunk 11]
```diff
@@ -477,9 +460,22 @@ struct Node : public Attributes<Node> {
...
       dropInput(i);
     inputs_.clear();
+    switch (alias_.kind()) {
+    case AliasKind::Multi:
+    case AliasKind::None:
+      // no-op
+      break;
+    case AliasKind::Select:
+    case AliasKind::Direct:
+      alias_ = Alias::none();
```
Comment (apaszke) [inline torch/csrc/jit/ir.h]: This part doesn't feel right to me. Say you have an in-place add constant. You want to rebase it in top of another part of the current graph so you do this:
```cpp
add->removeAllInputs();
add->addInput(x);
```
Now, the first call has removed the aliasing information BUT the function should still alias `x`. Aliasing information is a local property of the Function and it should only depend on parametrization (not on what is its input).

File: torch/csrc/jit/ir.h [hunk 12]
```diff
@@ -477,9 +460,22 @@ struct Node : public Attributes<Node> {
...
+      break;
+    }
```
Comment (apaszke) [inline torch/csrc/jit/ir.h]: This whole switch statement can be realized in 3 simple lines:
```cpp
if (alias_->kind() == AliasKind::Select || alias->kind() == AliasKind::Direct) {
    alias_ = Alias::none();
}
```

File: torch/csrc/jit/ir.h [hunk 13]
```diff
@@ -536,13 +534,14 @@ struct Node : public Attributes<Node> {
   Node* dropInput(size_t i) {
     JIT_ASSERT(i < inputs_.size());
     auto input_node = inputs_[i];
+    // NB: This is O(n)
```
Comment (apaszke) [inline torch/csrc/jit/ir.h]: I really think we shouldn't sprinkle code with comments like these. It's a characteristic of `findUseForInput`, not this function, so if someone changes it they will never think to update this comment. Misleading comments are worse than no comments

File: torch/csrc/jit/ir.h [hunk 14]
```diff
@@ -555,112 +554,28 @@ struct Node : public Attributes<Node> {
...
+    return new Node(g, kind());
   }
   // create a copy of all properties of Node s into this.
-  // subclasses should extend if they have additional informaiton to copy.
+  // subclasses should extend if they have additional information to copy.
   // 'this' will be allocated with s->allocNewInstance(g) so it should have
   // the same concrete type as 's'
   virtual void cloneFrom(Node * s) {
+    if (s->hasType()) setType(s->type());
+    setStage(s->stage());
```
Comment (apaszke) [inline torch/csrc/jit/ir.h]: I'm not sure if we should be cloning stages. There are legit situations where this is undesired, e.g. moving nodes to FusionGroups, which always have only one stage. `new_node_stage_` still applies when you clone and I would stick to that

File: torch/csrc/jit/ir.h [hunk 15]
```diff
@@ -671,13 +586,14 @@ friend struct Node;
   // as a Use object
   // also used as the beginning/end of the circular node list to avoid
   // having corner cases where the list is empty.
-  Node * output_;
+  // NB: initialized in constructor, and then never touched again
+  Node * output_ = nullptr;
```
Comment (apaszke) [inline torch/csrc/jit/ir.h]: you could have made it `const` and set it in the member list :wink:

File: torch/csrc/jit/ir.h [hunk 16]
```diff
@@ -698,17 +611,23 @@ friend struct Node;
...
 
   Node * addInput() {
-    Node* p = create(kParam);
-    inputs_.push_back(p);
-    return p;
+    return addInput(create(kParam));
+  }
+
+  Node * addInput(Node* n) {
+    inputs_.push_back(n);
```
Comment (apaszke) [inline torch/csrc/jit/ir.h]: assert is Param?

File: torch/csrc/jit/ir.h [hunk 17]
```diff
@@ -789,11 +709,15 @@ friend struct Node;
...
-    r->type_ = n->type_;
     r->cloneFrom(n);
     for(auto i : n->inputs()) {
       r->addInput(node_map(i));
     }
+    // NB: createClone is morally the same as create, but it doesn't
+    // call that function, so some logic gets duplicated
+    if (r->kind() == kParam) {
+      inputs_.push_back(r);
+    }
```
Comment (apaszke) [inline torch/csrc/jit/ir.h]: I think we should only allow cloning regular nodes and fail asserts if they're returns or params

File: torch/csrc/jit/python_ir.cpp [hunk 18]
```diff
@@ -23,7 +23,7 @@ void initPythonIRBindings(PyObject * module_) {
     .def("nodes",[](Graph &g) {
       return py::make_iterator(g.nodes().begin(),g.nodes().end());
     })
-    .GS(addInput)
+    .def("addInput",[](Graph &g) { return g.addInput(); })
```
Comment (apaszke) [inline torch/csrc/jit/python_ir.cpp:26]: ?

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 19]
```diff
@@ -196,30 +196,32 @@ struct GraphFuser {
...
+                                              JIT_ASSERT(chunk == chunk2);
+                                              return envFn[input];
+                                            });
+        insertAfter(ns, ns->input());
```
Comment (apaszke) [inline torch/csrc/jit/passes/graph_fuser.cpp]: Won't this reverse the selects in the topological order? I don't think we enforce this invariant anyhow, but it's a weird transformation

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 20]
```diff
@@ -229,7 +231,7 @@ struct GraphFuser {
   }
 
   // returns where to continue scanning
-  graph_node_list::iterator scanNode(Node * consumer) {
+  graph_node_list<Node>::iterator scanNode(Node * consumer) {
```
Comment (apaszke) [inline torch/csrc/jit/passes/graph_fuser.cpp]: if you really need these const lists can't you make them `const_graph_node_list` so we don't have to specify template args everywhere?

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 5 above)
Comment (apaszke) [reply torch/csrc/jit/passes/graph_fuser.cpp]: I agree with both of you haha. The older code was easier to read, but it's a bit of a code smell that we have to update it when we add aliasing etc

File: torch/csrc/autograd/python_function.cpp (hunk 8 above)
Comment (ezyang) [reply torch/csrc/autograd/python_function.cpp]: ~~Do you mean `VariableImpl*`? Because `Variable` is a PIMPLy thing, `Variable*` isn't really a coherent concept?~~ I'm confused, I thought I had an unordered_map but I don't, `Variable*` is fine.

File: torch/csrc/jit/attributes.h (hunk 9 above)
Comment (ezyang) [reply torch/csrc/jit/attributes.h:156]: I think what happened was, Zach originally implemented it with unordered_map, but the test outputs kept wobbling (remember we print the attributes), so then it turned into a vector. Maybe we should do something more efficient here.

File: torch/csrc/jit/ir.h (hunk 10 above)
Comment (ezyang) [reply torch/csrc/jit/ir.h]: The problem is that the rules for propagation of types versus aliases are different.

Suppose that I have some variable `x : Float(32, 32)`. If I bind this to another variable `y`, the type stays the same.

Aliasing, however, is a structural property of the *computation itself*; it's only attached to nodes (which denote values) because we don't actually have anywhere to put it otherwise (darn you, lack of a separate Operator struct!) Logically, if I have computation `View`, I might say that its output aliases its input. But if I call the output of `View` variable `x`, and I say its alias is `SelectAlias 0`, if I then bind `x` to `y`, it's totally nonsense to say that `y` is `SelectAlias 0`.

To be fair, I'm kind of not happy about having to attach aliasing informations to nodes (i.e. values). It's denotationally incoherent. But we refactored away from having a separate notion of "operation", and unless we undo that refactor, there's no where else to put the information.

By the way, why did I put the alias information in the select nodes, rather than in the computation node itself? No particularly good reason, except for consistency with how types work (which is perhaps why you asked this question in the first place.) Perhaps this is a foolish consistency!

File: torch/csrc/jit/ir.h (hunk 12 above)
Comment (ezyang) [reply torch/csrc/jit/ir.h]: But then you don't get exhaustivity checking :)

File: torch/csrc/jit/ir.h (hunk 11 above)
Comment (ezyang) [reply torch/csrc/jit/ir.h]: I agree with you!

The crux of the matter is, semantically, what do we mean by `removeAllInputs`; i.e., under what situations would you like to call `removeAllInputs`? At the moment, it is used exclusively when we are destroying a node, to make sure that we update the uses of all of its inputs. This is legitimate use case, although it is a bit specific to the implementation, which might suggest that removeAllInputs should be a private method.

But you suggested we could do this rebase. I actually think that you shouldn't implement it the way you suggested. Here's another way to do it:

```
new_add = g->createClone(add, [](Node *x) { return /* mapping to new inputs */; });
// insert new_add where you want it
add->replaceAllUsesWith(new_add)
add->destroy()
```

Despite the extra allocations, I like this a lot better. Why? Because there is never a point in time outside of internal code in Graph/Node's implementation that an add/new_add in the graph is malformed.

Now, we have never agreed that this was actually a property we want to preserve. Certainly it is true that we don't expect `graph->lint()` to pass in-between calls to the graph: we don't expect nodes which are not actually placed in the graph to be well-formed. But I think there is a subset of lint properties which we can and *should* expect to be held every operation. And if there is some client code which needs to violate it, we should be very clear about this, and it should call lint as soon as the invariants are restored.

**To summarize.** Let's make `removeAllInputs` and other input twiddling functions legal *only* when a node is not attached to the graph. These functions would then make no modifications to aliasing information; they're purely convenience mutators to help you swizzle inputs into the correct form before you put them into the graph.

File: torch/csrc/jit/python_ir.cpp (hunk 18 above)
Comment (ezyang) [reply torch/csrc/jit/python_ir.cpp:26]: addInput got an extra overload so pybind11 needs help picking the correct overload.

File: torch/csrc/jit/ir.h (hunk 14 above)
Comment (ezyang) [reply torch/csrc/jit/ir.h]: On the other hand, if you were writing a simplifier, failing to clone stages would result in your new nodes always getting allocated to the latest stage.

What is our design for stages? My original design for stages was that they should be implicit, not explicit concept. That is to say, inputs would have explicit stages, but then every other computation `f(x1,x2,x3)` would have the max stage of all of its inputs. In this model, we'd always move computations to the earliest stage they could execute. The problem with this model is you destroy knowledge about the original staging, which may have been done the way it was to reduce working set size across forward-backward.

Given that we're now in the business of explicitly maintaining stages, perhaps a refactor that would make sense is to encode stages structurally into the graph data structure. So now, rather than having a single doubly-linked list of nodes, we have a doubly-linked list *per stage*. We can then eliminate `new_node_stage_` for the graph as a whole, because now the stage assigned to a node is based on *where you put it*, which is a coherent concept. This also helps us enforce the fact that we already require stages to be lined up in topological order. (An alternate way, less intrusive way, to do this refactoring is to keep a single linked list, but compute stage upon insertion based on the reference node it was inserted with.)

File: torch/csrc/jit/ir.h (hunk 17 above)
Comment (ezyang) [reply torch/csrc/jit/ir.h]: Oops, I needed to refactor this some more. There shouldn't be a `push_back` here.

**Comment by ezyang** [comment]
(If anyone is wondering, "Why am I so obsessed with IR invariants? Why can't we just have a bag of useful manipulation utility functions, and trust users to use them correctly?" just shout, I'll be more than happy to explain :)

File: torch/csrc/jit/attributes.h (hunk 9 above)
Comment (zdevito) [reply torch/csrc/jit/attributes.h:156]: (Also the vector is going to be faster, there are like 3 attributes max and the cross over between vector and unorder map perf is probably closer to 100 elements than 3)

File: torch/csrc/jit/ir.h (hunk 15 above)
Comment (ezyang) [reply torch/csrc/jit/ir.h]: Oh all right ^^"

**Comment by ezyang** [comment]
See new FIXUP commits.

File: torch/csrc/jit/python_ir.cpp (hunk 18 above)
Comment (apaszke) [reply torch/csrc/jit/python_ir.cpp:26]: ah, I think people usually just add `static_cast`s to the appropriate type if it's ambiguous

File: torch/csrc/jit/ir.h (hunk 10 above)
Comment (apaszke) [reply torch/csrc/jit/ir.h]: What does "bind" mean in your explanation?

File: torch/csrc/jit/ir.h (hunk 11 above)
Comment (apaszke) [reply torch/csrc/jit/ir.h]: I never said it's the best way to implement rebase, but it's definitely _**a way**_ and it looks perfectly valid, so it would be very very hard to catch in code review.

What do you define by "graph attachment"? Nodes are always associated with some graph (that's how their lifetime is defined). Do you mean that they're not in the list that defines our topological ordering? It feels weird to say that this function is legal only then, because aliasing has absolutely no connection to the ordering we pick.

Comment (ezyang) [reply torch/csrc/jit/ir.h]: Right, this is why I suggest making `removeAllInputs` error if it's run on a node in a graph, and we'll have it not update alias information. It's also somewhat difficult to catch when people mess up alias info on graph transformations, so I also want to make it hard to mess this up.

> What do you define by "graph attachment"? Nodes are always associated with some graph (that's how their lifetime is defined). Do you mean that they're not in the list that defines our topological ordering?

Yes, that's what I mean. The point of the restriction is to make you less likely to call `removeInputs` in a context where it is semantically incorrect and trashing some analysis information you care about.

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 5 above)
Comment (zdevito) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Right neither is really a clone. I think the clearest thing would be to have no use of clone here. Instead use `copyMetadata` when appropriate.

File: torch/csrc/jit/ir.h (hunk 10 above)
Comment (ezyang) [reply torch/csrc/jit/ir.h]: "bind" = as in a let-binding / single assignment. Like, let x = y. I know we don't "support" this natively. But it's good to think about.

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 5 above)
Comment (ezyang) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Want to double check my understanding of the code. The comment says:

```
  // replace a,b = chunk(op(x,y,z)) with:
  // x0,x1 = chunk(x) (x0 has a's type, x1 has b's type)
  // y0,y1 = chunk(y) (y0 has a's type, y1 has b's type)
  // z0,z1 = chunk(z) (z0 has a's type, z1 has b's type)
  // a = op(x0,y0,z0) (a,b have their same size but are now contiguous)
  // b = op(x1,y1,x1)
```

Suppose that op is a function which involves two types (e.g., "and :: float, float -> bool"). Then a, b are bool tensors, and x is a float tensor. So chunk of x should not be a, b.

There is an interpretation that is consistent with the code (and the operators which are currently supported): simple_mappable operators are *always* return outputs which are the same type (this is not documented anywhere, it just happens to be true for the operators we have implemented fusion for.) Adding to the confusion, the comment implies that a and b have different types; this is true in a formal sense, but it is really only their *sizes* which differ.

(Also, while thinking about this, I realized that the fuser doesn't detect if broadcasting has occurred and bail out...)

Comment (zdevito) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Yes, the type here is purely talking about their sizes. If op can change the type, then we need more logic here. However, when this happens, we may want to just go to an approach based on semantic analysis for determining types. Once we have enough optimizations, constantly doing type propagation at the site of the modification will get error prone.

Comment (ezyang) [reply torch/csrc/jit/passes/graph_fuser.cpp]: I ended up writing the logic for this particular case correctly. I agree about the semantic analysis but that's another chunk of design work that would need to be done. (Let's write a bidirectional typechecker!!)

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 19 above)
Comment (ezyang) [reply torch/csrc/jit/passes/graph_fuser.cpp]: I fixed this in my latest commit.

File: torch/csrc/jit/ir.h (hunk 11 above)
Comment (ezyang) [reply torch/csrc/jit/ir.h]: CC @zdevito here.

**Comment by ezyang** [comment]
The two unfinished comments are:

* https://github.com/ezyang/pytorch/pull/229#discussion_r139161542 (removeInputs semantics with alias information)
* https://github.com/ezyang/pytorch/pull/229/files/310daead108806e34b99428d3a81b40aae4bb670#diff-c7fa0d815e308a9fc70b52c9a049a8a3 (on the cloning of stages)

**Comment by ezyang** [comment]
Since we're dropping aliasing now, review up to fac237eb880768d3521f102a67f743c7fc91f7f6, but there are other patches on top which need to be moved backwards.

**Comment by ezyang** [comment]
OK, aliasing has been excised.

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 21]
```diff
@@ -196,33 +211,65 @@ struct GraphFuser {
...
-        ns->setType(sel->typeOption());
-        insertAfter(ns,c);
-        return ns;
-      });
-      if(sel->hasType()) {
-        new_output->setType(sel->type()->cast<TensorType>()->contiguous());
+    // apply the op to each chunk of the chunked operands,
+    // and then rewrite the graph to use them!
+    // NB: as we replace/remove the selects the use list changes, so copy it first
+    auto chunk_outputs = chunk->outputs();
```
Comment (zdevito) [inline torch/csrc/jit/passes/graph_fuser.cpp]: This is unused.

**Comment by zdevito** [comment]
Looks good now. I am going to merge as soon as the rebase I just did passes the tests.
